quinn-h3/src/client.rs

```diff
@@ -0,0 +1,286 @@
+                                _ => unreachable!(),
+                            };
+                        }
+                        _ => return Err(Error::peer("first frame is not headers")),
+                    },
+                },
+                SendRequestState::Decoding(ref mut frame) => {
+                    let stream_id = self.stream_id.ok_or(Error::Internal("Stream id is none"))?;
+                    let result = {
+                        let conn = &mut self.conn.h3.lock().unwrap().inner;
+                        conn.decode_header(&stream_id, frame)
+                    };
+
+                    match result {
+                        Ok(None) => return Ok(Async::NotReady),
+                        Ok(Some(decoded)) => {
+                            self.state = SendRequestState::Ready(decoded);
+                        }
+                        Err(e) => {
+                            return Err(Error::peer(format!("decoding header failed: {:?}", e)))
+                        }
+                    }
+                }
+                SendRequestState::Ready(_) => {
+                    match mem::replace(&mut self.state, SendRequestState::Finished) {
+                        SendRequestState::Ready(h) => {
+                            return Ok(Async::Ready(RecvResponse::build(
+                                h,
+                                try_take(&mut self.recv, "Recv is none")?,
+                                try_take(&mut self.stream_id, "stream is none")?,
+                                self.conn.clone(),
+                            )?));
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                _ => return Err(Error::Poll),
+            }
+        }
+    }
+}
+
+pub struct RecvResponse {
+    response: Response<()>,
+    recv: FrameStream<RecvStream>,
+    stream_id: StreamId,
+    conn: ConnectionRef,
+}
+
+impl RecvResponse {
+    fn build(
+        header: Header,
+        recv: FrameStream<RecvStream>,
+        stream_id: StreamId,
+        conn: ConnectionRef,
+    ) -> Result<Self, Error> {
+        let (status, headers) = header.into_response_parts()?;
+        let mut response = Response::builder();
+        response.status(status);
+        response.version(http::version::Version::HTTP_2); // TODO change once available
+        *response
+            .headers_mut()
+            .ok_or(Error::peer("invalid response"))? = headers;
+
+        Ok(Self {
+            recv,
+            conn,
+            stream_id: stream_id,
+            response: response
+                .body(())
+                .or(Err(Error::Internal("failed to build response")))?,
+        })
+    }
+
+    pub fn response<'a>(&'a self) -> &'a Response<()> {
+        &self.response
+    }
+
+    pub fn body(self) -> RecvBody {
+        RecvBody::with_capacity(self.recv, 10240, 1024000, self.conn.clone(), self.stream_id)
```

Comment:
Those static values aren't to be left like this. I'm not sure where to store them, and how to let the user set them:
- They could vanish on body streaming implementation. If we decide to replace the current API with streamin only.
- use setters as member of `RecvBody` like this: 
```rust
response.set_max_capacity(n);
response.set_capacity(m);
response.body().and_then(...)
```

Reply:
I think we should have two modes of operation for bodies: (a) an all-at-once mode that keeps a growable buffer internally and returns the entire response body at once to the caller, and (b) a streaming mode that has a limited fixed-size buffer whose size is configured for the `Connection` on construction.